modules/common-fs2/src/main/scala/com/snowplowanalytics/snowplow/enrich/common/fs2/Enrich.scala

```diff
@@ -64,33 +68,40 @@ object Enrich {
    * [[Assets.State.make]] downloads assets for the first time unconditionally during
    * [[Environment]] initialisation, then if `assetsUpdatePeriod` has been specified -
    * they'll be refreshed periodically by [[Assets.updateStream]]
-   *
-   * @param ordered indicates whether the events should be processed ordered or not
    */
-  def run[F[_]: Concurrent: ContextShift: Clock: Parallel, A](env: Environment[F, A], ordered: Boolean): Stream[F, Unit] = {
+  def run[F[_]: Concurrent: ContextShift: Clock: Parallel: Timer, A](env: Environment[F, A]): Stream[F, Unit] = {
     val registry: F[EnrichmentRegistry[F]] = env.enrichments.get.map(_.registry)
     val enrich: Enrich[F] = {
       implicit val rl: RegistryLookup[F] = env.registryLookup
       enrichWith[F](registry, env.igluClient, env.sentry, env.processor)
     }
 
-    val enrichPipe: Pipe[F, (A, Array[Byte]), (A, Result)] =
-      in => {
-        val parallel =
-          if (ordered)
-            in.parEvalMap[F, (A, Result)](env.streamsSettings.concurrency.enrichment) _
-          else
-            in.parEvalMapUnordered[F, (A, Result)](env.streamsSettings.concurrency.enrichment) _
-        parallel { case (orig, bytes) => enrich(bytes).map((orig, _)) }
-      }
-
-    env.source
-      .pauseWhen(env.pauseEnrich)
-      .evalTap(_ => env.metrics.rawCount)
-      .fproduct(env.getPayload)
-      .through(enrichPipe)
-      .through(sinkResult(sinkOne(env), env.metrics.enrichLatency, env.streamsSettings.concurrency.output))
-      .through(env.checkpointer)
+    val enriched =
+      env.source.chunks
+        .prefetch
+        .evalTap(chunk => Logger[F].debug(s"${DateTime.now().toString()} - Starting to process chunk of size ${chunk.size}"))
+        .pauseWhen(env.pauseEnrich)
```

Comment:
Yes I remember our discussion I was planning to address it in the other PR, but I'll address it now.

Reply:
I've rewrote the assets mechanism so that [we use](https://github.com/snowplow/enrich/blob/feature/enrich-kinesis/modules/common-fs2/src/main/scala/com/snowplowanalytics/snowplow/enrich/common/fs2/Environment.scala#L80) a `Semaphore` instead of a `SignallingRef`.

Its permit is shared between [enrich](https://github.com/snowplow/enrich/blob/feature/enrich-kinesis/modules/common-fs2/src/main/scala/com/snowplowanalytics/snowplow/enrich/common/fs2/Enrich.scala#L85) and [updating the assets](https://github.com/snowplow/enrich/blob/feature/enrich-kinesis/modules/common-fs2/src/main/scala/com/snowplowanalytics/snowplow/enrich/common/fs2/Assets.scala#L169).

The new logic for the refresh mechanism is to:
1. [Download all the assets](https://github.com/snowplow/enrich/blob/feature/enrich-kinesis/modules/common-fs2/src/main/scala/com/snowplowanalytics/snowplow/enrich/common/fs2/Assets.scala#L163) locally into a temporary location, without stopping the enriching
2. [Checks if some assets have changed](https://github.com/snowplow/enrich/blob/feature/enrich-kinesis/modules/common-fs2/src/main/scala/com/snowplowanalytics/snowplow/enrich/common/fs2/Assets.scala#L165) by comparing the hashes with the ones stored in the `Assets.State`
3. If there has been an update, then stop enriching and [update the assets](https://github.com/snowplow/enrich/blob/feature/enrich-kinesis/modules/common-fs2/src/main/scala/com/snowplowanalytics/snowplow/enrich/common/fs2/Assets.scala#L170)

There are 2 new unit tests :
1. [One](https://github.com/snowplow/enrich/blob/feature/enrich-kinesis/modules/common-fs2/src/test/scala/com/snowplowanalytics/snowplow/enrich/common/fs2/AssetsSpec.scala#L51-L85) that checks assets initialization
2. [One](https://github.com/snowplow/enrich/blob/feature/enrich-kinesis/modules/common-fs2/src/test/scala/com/snowplowanalytics/snowplow/enrich/common/fs2/AssetsSpec.scala#L110-L143) that inits the state with v1 of an asset, runs the update stream where v2 is now available, and checks that the asset has been updating to v2 locally

This can look weird that we are passing the [list of assets](https://github.com/snowplow/enrich/blob/feature/enrich-kinesis/modules/common-fs2/src/main/scala/com/snowplowanalytics/snowplow/enrich/common/fs2/Assets.scala#L156) to `updateStream` whereas we can get the information [from the enrichments](https://github.com/snowplow/enrich/blob/feature/enrich-kinesis/modules/common-fs2/src/main/scala/com/snowplowanalytics/snowplow/enrich/common/fs2/Assets.scala#L154). I had removed `List[Asset]` to then realize that with previous unit tests implementation it was not possible so I reverted the changes. Now that I rewrote the tests I think that we can make it work. Please let me know if you think that we should do it (not very important I think, just a neat), otherwise I'll leave it as is.